{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import datetime\n",
    "from subprocess import call\n",
    "import subprocess\n",
    "import glob\n",
    "import djPyBio as DJ\n",
    "from djPyBio import Common as CM\n",
    "import argparse\n",
    "\n",
    "import pandas as pd\n",
    "import csv\n",
    "import copy \n",
    "import pybedtools as pbt\n",
    "import ciepy\n",
    "import cardipspy as cpy\n",
    "import itertools\n",
    "import tempfile\n",
    "import six\n",
    "import networkx as nx\n",
    "from scipy.stats import mode\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.max_rows', 100)\n",
    "from mpl_toolkits.axes_grid1 import  make_axes_locatable\n",
    "import datetime\n",
    "\n",
    "from scipy.stats import mode\n",
    "import scipy.stats as stats\n",
    "\n",
    "from collections import Counter\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_cnvs(cnvs, info):\n",
    "    \"\"\"Combine the list of CNVs cnvs into a single CNV of form CNV_{chrom}_{start}_{end}.\"\"\"\n",
    "    return 'CNV_{}_{}_{}'.format(info.loc[cnvs[0], 'Chr'], \n",
    "                                 info.loc[cnvs, 'Start'].min(), \n",
    "                                 info.loc[cnvs, 'End'].max())\n",
    "\n",
    "def merge_cnvs(a, b, info, cns):\n",
    "    \"\"\"Return boolean indicating whether CNVs a and b should be merged.\"\"\"\n",
    "    return ((abs(info.loc[a, 'order'] - info.loc[b, 'order']) == 1) and\n",
    "            ((cns.loc[a] - cns.loc[b]).abs().mean() < 0.5))\n",
    "\n",
    "\n",
    "# In[38]:\n",
    "\n",
    "def compare_lists(l1, l2):\n",
    "    count = 0\n",
    "    for i1, i2 in zip(l1, l2):\n",
    "        if i1 != i2:\n",
    "            count +=1\n",
    "    return count\n",
    "def calculate_absolute_mean_diff(a1, a2):\n",
    "    \"\"\"calculate the absolute mean difference of elements in 2 equal length arrays\"\"\"\n",
    "    diff = np.array(a1) - np.array(a2)\n",
    "    diff_abs = np.absolute(diff)\n",
    "    mean_diff = diff_abs.mean()\n",
    "    return mean_diff\n",
    "\n",
    "\n",
    "# In[76]:\n",
    "\n",
    "def compare_sites(ind1, ind2, cns_t, samples,samples_nmode1, samples_nmode2, cnmode1, cnmode2, samples_lq1, samples_lq2, subtract_lq = True):\n",
    "    \"\"\" calculate the pearson corr of sites in the intersection, calculate the number of differences among non-mode samples \"\"\"\n",
    "    data = []\n",
    "\n",
    "      \n",
    "    ids = [ind1, ind2]\n",
    "    ids_mod = [\"_\".join(i.split('_')[:-1]) for i in ids]\n",
    "    \n",
    "    \n",
    "    cns_1 = cns_t[ind1].to_dict()\n",
    "    cns_2 = cns_t[ind2].to_dict()\n",
    "    \n",
    "        \n",
    "    samples_nmode1 = [i for i in samples if cns_1[i] != cnmode1]\n",
    "    samples_nmode2 = [i for i in samples if cns_2[i] != cnmode2]\n",
    "    samples_to_compare = set(samples_nmode1 + samples_nmode2)\n",
    "    if subtract_lq:\n",
    "        # exclude LQ samps\n",
    "        samples_to_exclude = set(samples_lq1 + samples_lq2)\n",
    "        samples_to_compare_nmode = list(samples_to_compare.difference(samples_to_exclude))\n",
    "        \n",
    "        samples_to_compare_nmode = list(samples_to_compare.difference(samples_to_exclude))\n",
    "        samples_to_compare_corr = list(set(samples).difference(samples_to_exclude))\n",
    "    \n",
    "    else: \n",
    "        samples_to_compare_corr = samples\n",
    "        samples_nmode1 = [i for i in samples if cns_1[i] != cnmode1]\n",
    "        samples_nmode2 = [i for i in samples if cns_2[i] != cnmode2]\n",
    "        samples_to_compare_nmode = list(set(samples_nmode1 + samples_nmode2))\n",
    "        samples_to_exclude = []\n",
    "    \n",
    "    cns_nmode_1 = [int(cns_1[i]) for i in samples_to_compare_nmode]\n",
    "    cns_nmode_2 = [int(cns_2[i]) for i in samples_to_compare_nmode]\n",
    "\n",
    "    cns_corr_1 = [int(cns_1[i]) for i in samples_to_compare_corr]\n",
    "    cns_corr_2 = [int(cns_2[i]) for i in samples_to_compare_corr]\n",
    "    \n",
    "    allele_dist1 = dict(Counter(cns_corr_1))\n",
    "    allele_dist2 = dict(Counter(cns_corr_2))\n",
    "    \n",
    "    corr_coef = stats.pearsonr(cns_corr_1, cns_corr_2)[0]\n",
    "        \n",
    "    nsamp = len(samples_to_compare_nmode)\n",
    "    nsamp_pass = len(samples_to_compare_corr)\n",
    "\n",
    "    num_diff = compare_lists(cns_nmode_1, cns_nmode_2)\n",
    "    alleles = set(cns_corr_1 + cns_corr_2)\n",
    "    \n",
    "    alleles1 = set(cns_corr_1)\n",
    "    num_alleles1 = len(alleles1)\n",
    "    alleles2 = set(cns_corr_2)\n",
    "    num_alleles2 = len(alleles2)\n",
    "\n",
    "    mean_diff_all = calculate_absolute_mean_diff(cns_corr_1, cns_corr_2)\n",
    "    mean_diff_nmode = calculate_absolute_mean_diff(cns_nmode_1, cns_nmode_2)\n",
    "    \n",
    "    exact_match = (cns_corr_1 == cns_corr_2)\n",
    "\n",
    "    \n",
    "    try:\n",
    "        perc_diff = num_diff/nsamp\n",
    "    except:\n",
    "#         print nsamp, 'nsamp is zero'\n",
    "        perc_diff = 0\n",
    "\n",
    "    out = [ind1, ind2, corr_coef, num_diff, nsamp, \n",
    "           perc_diff, samples_to_compare_nmode, samples_to_compare_corr, \n",
    "           list(samples_to_exclude), nsamp_pass, exact_match, mean_diff_all, mean_diff_nmode, \n",
    "           alleles1, alleles2, num_alleles1, num_alleles2, allele_dist1, allele_dist2]\n",
    "\n",
    "    return out\n",
    "\n",
    "\n",
    "def split_id_to_coord(ID):\n",
    "    spl = ID.split('_')\n",
    "    chrom = spl[1]\n",
    "    start = int(spl[2])\n",
    "    end = int(spl[3])\n",
    "    return chrom, start, end\n",
    "\n",
    "\n",
    "# In[41]:\n",
    "\n",
    "def compute_dist(id1, id2):\n",
    "    chrom1, start1, end1 = split_id_to_coord(id1)\n",
    "    chrom2, start2, end2 = split_id_to_coord(id2)\n",
    "    dist = start2 - end1\n",
    "    return dist\n",
    "    \n",
    "\n",
    "\n",
    "# In[65]:\n",
    "\n",
    "def collect_data_adjacent_sites(info, cns_t, samples, cn_mode_col = 'cn_mode', subtract_lq=True):\n",
    "    \n",
    "    data = []\n",
    "\n",
    "    for chrom, df in info.groupby('Chr'):\n",
    "        df = df.sort_values('order')\n",
    "        inds = df.index.tolist()\n",
    "        diff_mode_uuids = info.diff_mode_uuids.to_dict()\n",
    "        mode_cn_all = info[cn_mode_col].to_dict()\n",
    "        lq_uuids = info.lq_samps.to_dict()\n",
    "        cnv_classes = info.cnv_class.to_dict()\n",
    "\n",
    "\n",
    "#         if chrom not in ['X', 'Y']:\n",
    "        max_range = len(inds)-1 \n",
    "        for i in range(0, max_range):\n",
    "            if i < max_range:\n",
    "\n",
    "                ind1 = inds[i]\n",
    "                ind2 = inds[i+1]\n",
    "                pair_ind = '-'.join([ind1, ind2])\n",
    "\n",
    "                distance_between = compute_dist(ind1, ind2)\n",
    "                if distance_between < 0:\n",
    "                    absolute_dist = 0\n",
    "                else:\n",
    "                    absolute_dist = distance_between\n",
    "\n",
    "                cnv_class1 = cnv_classes[ind1]\n",
    "                cnv_class2 = cnv_classes[ind2]\n",
    "\n",
    "                diff_uuids1 = diff_mode_uuids[ind1]\n",
    "                diff_uuids2 = diff_mode_uuids[ind2]\n",
    "                mode_cn1 = mode_cn_all[ind1]\n",
    "                mode_cn2 = mode_cn_all[ind2]\n",
    "                lq_uuids1 = lq_uuids[ind1]\n",
    "                lq_uuids2 = lq_uuids[ind2]\n",
    "                pair = [ind1,ind2]\n",
    "\n",
    "                num_diff1 = len(diff_uuids1)\n",
    "                num_diff2 = len(diff_uuids2)\n",
    "\n",
    "                comp = compare_sites(ind1, ind2, cns_t, samples, \n",
    "                                     diff_uuids1, diff_uuids2, mode_cn1, mode_cn2, \n",
    "                                     lq_uuids1, lq_uuids2, subtract_lq=subtract_lq)\n",
    "\n",
    "                comp = comp + [pair, mode_cn1, mode_cn2, cnv_class1, cnv_class2, distance_between, \n",
    "                               absolute_dist, pair_ind, num_diff1, num_diff2, diff_uuids1, diff_uuids2,\n",
    "                              chrom]\n",
    "                data.append(comp)\n",
    "\n",
    "    df = pd.DataFrame(data, columns=['ID1', 'ID2', 'corr_coef',\n",
    "                                     'num_diff', 'num_non_mode', 'percent_non_mode_diff', \n",
    "                                     'samps_to_compare_nmode', 'samps_to_compare_corr',\n",
    "                                     'samps_to_exclude', 'num_pass', 'exact_cn_match', \n",
    "                                     'mean_cn_diff_all', 'mean_cn_diff_nmode','alleles1', 'alleles2', \n",
    "                                     'num_alleles1', 'num_alleles2','allele_dist1', 'allele_dist2','pair', \n",
    "                                     'mode_cn1', 'mode_cn2', 'cnv_class1', 'cnv_class2', \n",
    "                                     'distance_between', 'distance_between_mod', 'cat_pair', \n",
    "                                     'num_diff1', 'num_diff2', 'diff_uuids1', 'diff_uuids2', 'chrom'])\n",
    "    \n",
    "    df.index = df.cat_pair\n",
    "    return df\n",
    "\n",
    "\n",
    "def prep_info(df):\n",
    "    df = df.copy()\n",
    "    df = df.sort_values(['Chr', 'Start', \"End\"])\n",
    "    df['order'] = range(0, df.shape[0])\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "def prep_cns(df, info):\n",
    "    df = df.copy()\n",
    "    df = df.reindex(info.index)\n",
    "    return df \n",
    "\n",
    "\n",
    "\n",
    "def annotate_passing(df, thresh):\n",
    "    df = df.copy()\n",
    "    inds = df[(df.corr_coef > 0.9) & (df.percent_non_mode_diff <= 0.2) & (df.mean_cn_diff_all < 0.5) &\n",
    "              (df.mean_cn_diff_nmode < 0.5) & (df.distance_between_mod < thresh)].index.tolist()\n",
    "    \n",
    "\n",
    "    new_size = df.shape[0]\n",
    "    \n",
    "    \n",
    "    df['passing_criteria'] = False\n",
    "    df.loc[inds, 'passing_criteria'] = True\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "def prep_adj_sites(adj_sites_prestich, thresh = 5000):  \n",
    "    adj_sites_prestich = adj_sites_prestich.copy()\n",
    "    adj_sites_prestich['matching_mode'] = (adj_sites_prestich.mode_cn1 == adj_sites_prestich.mode_cn2)\n",
    "#     adj_sites_plot = adj_sites_prestich[(adj_sites_prestich.matching_mode == True) & \n",
    "#                                     (adj_sites_prestich.num_alleles1 > 1) & \n",
    "#                                         (adj_sites_prestich.num_alleles2 > 1)].copy()\n",
    "    adj_sites_prestich['log_dist'] = np.log10(adj_sites_prestich.distance_between_mod + 1)\n",
    "    adj_sites_prestich = adj_sites_prestich.pipe(annotate_passing, thresh)\n",
    "    return adj_sites_prestich\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_corr_at_samples(ind1, ind2, samples, cns_t):\n",
    "    cns_1 = cns_t[ind1].to_dict()\n",
    "    cns_2 = cns_t[ind2].to_dict()\n",
    "    try:\n",
    "        cns_corr1 = [int(cns_1[i]) for i in samples]\n",
    "        cns_corr2 = [int(cns_2[i]) for i in samples]\n",
    "        corr_coef = stats.pearsonr(cns_corr1, cns_corr2)[0]\n",
    "    except:\n",
    "        print samples\n",
    "    return corr_coef\n",
    "\n",
    "\n",
    "def gather_consensus_X(adj_sites_all_males, adj_sites_all_females, adj_sites, cns_t):\n",
    "    inds = adj_sites_all_males.index.tolist()\n",
    "    assert (inds == adj_sites_all_females.index.tolist())\n",
    "    \n",
    "    num_non_mode_males = adj_sites_all_males.num_non_mode.tolist()\n",
    "    num_non_mode_females = adj_sites_all_females.num_non_mode.tolist()\n",
    "    \n",
    "    ndiff_male = adj_sites_all_males.num_diff.tolist()\n",
    "    ndiff_female = adj_sites_all_females.num_diff.tolist()\n",
    "    \n",
    "    passing_b_males = adj_sites_all_males.passing_criteria.tolist()\n",
    "    passing_b_females = adj_sites_all_females.passing_criteria.tolist()\n",
    "    \n",
    "    samples_nm_males = adj_sites_all_males.samps_to_compare_corr.tolist()\n",
    "    samples_nm_females = adj_sites_all_females.samps_to_compare_corr.tolist()\n",
    "    union_samples = [list(set(i1 + i2)) for i1, i2 in zip(samples_nm_males, samples_nm_females)]\n",
    "   \n",
    "    corr_males = adj_sites_all_males.corr_coef.tolist()\n",
    "    corr_females = adj_sites_all_females.corr_coef.tolist()\n",
    "    dist = adj_sites_all_males.distance_between_mod.tolist()\n",
    "    \n",
    "    out = []\n",
    "    for ind, nm_male, nm_female, pass_b_male, pass_b_female, cm, cf, ndm, ndf, d, us in zip(inds,\n",
    "                                                                                            num_non_mode_males,\n",
    "                                                                           num_non_mode_females,\n",
    "                                                                           passing_b_males, passing_b_females, \n",
    "                                                                           corr_males, corr_females,\n",
    "                                                                           ndiff_male, ndiff_female, \n",
    "                                                                           dist, union_samples):\n",
    "        \n",
    "        \n",
    "        ind1, ind2 = ind.split('-')\n",
    "        chrom = ind.split('_')[1]\n",
    "        \n",
    "        if chrom == 'X':\n",
    "\n",
    "            if [nm_male, nm_female] == [0, 0]:\n",
    "                out.append([ind, False])\n",
    "\n",
    "            elif (nm_male == 0) & (nm_female > 0):\n",
    "                out.append([ind, pass_b_female])\n",
    "\n",
    "            elif (nm_male > 0) & (nm_female == 0):\n",
    "                out.append([ind, pass_b_male])\n",
    "\n",
    "            elif (nm_male > 0) & (nm_female > 0):\n",
    "                if all([pass_b_male, pass_b_female]):\n",
    "                    out.append([ind, True])\n",
    "                else:\n",
    "                    out.append([ind,False])\n",
    "                    \n",
    "            else:\n",
    "                print \"didn't account for all scenarios\"\n",
    "                break\n",
    "                \n",
    "    df = pd.DataFrame(out, columns=['cat_pair', 'passing_consensus'])\n",
    "    df.index = df.cat_pair\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_x_chrom(df):\n",
    "    df = df.copy()\n",
    "    df = df[df.chrom=='X']\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compile_stitch_clusters(adj_sites, info):\n",
    "    \n",
    "    out = dict()\n",
    "    adj_sites['order'] = range(0, adj_sites.shape[0])\n",
    "    \n",
    "    for chrom, df in adj_sites.groupby('chrom'):\n",
    "        # ensure order maintained\n",
    "        df = df.sort_values('order')\n",
    "        combine = df.passing_criteria_adjusted.tolist()\n",
    "        last_cnv = df.ID2.tolist()[-1]\n",
    "        cnvs = df.ID1.tolist() + [last_cnv]\n",
    "        \n",
    "        i = 0    \n",
    "        to_combine = [cnvs[0]]\n",
    "        while i < len(cnvs) - 1:\n",
    "            if combine[i]:\n",
    "                to_combine.append(cnvs[i + 1])\n",
    "            else:\n",
    "                if len(to_combine) > 1:\n",
    "                    out[combine_cnvs(to_combine, info)] = to_combine\n",
    "                to_combine = [cnvs[i + 1]]\n",
    "            i += 1\n",
    "        # if we end on a True - combine what is left\n",
    "        if len(to_combine) > 1:\n",
    "            out[combine_cnvs(to_combine, info)] = to_combine\n",
    "\n",
    "    return out\n",
    "    \n",
    "\n",
    "def characterize_cnv_classes(x):\n",
    "    Types = []\n",
    "    out = 'none'\n",
    "    if ('DUP' in x) and ('mCNV' in x) and ('DEL' not in x):\n",
    "        out = 'DUP,mCNV'\n",
    "    if ('DUP' in x )and ('DEL' not in x) and ('mCNV' not in x):\n",
    "        out = 'DUP,DUP'\n",
    "    if ('DEL' in x) and ('DUP' not in x) and ('mCNV' not in x):\n",
    "        out = 'DEL,DEL'\n",
    "    if ('DEL' in x) and ('mCNV' in x) and ('DUP' not in x):\n",
    "        out = 'DEL,mCNV'\n",
    "    if ('DEL' in x) and ('mCNV' in x) and ('DUP' in x):\n",
    "        out = 'DEL,DUP,mCNV'\n",
    "    if ('DEL' in x) and ('DUP' in x) and ('mCNV' not in x):\n",
    "        out = 'DUP,DEL'\n",
    "    if ('mCNV' in x) and ('DUP' not in x) and ('DEL' not in x):\n",
    "        out = 'mCNV,mCNV'\n",
    "        \n",
    "    return out\n",
    "\n",
    "def lambda_stitch_site_in_cnvs_merged(x):\n",
    "    merged_cnvs = x['cnvs_merged']\n",
    "    stitched_cnv = x['stitched_cnv_site_ID']\n",
    "    \n",
    "    for c in merged_cnvs:\n",
    "        if \"_\".join(c.split('_')[:-1]) == stitched_cnv:\n",
    "            return [True, c, stitched_cnv]\n",
    "    return [False, False, stitched_cnv]\n",
    "\n",
    "\n",
    "\n",
    "def lambda_add_stitched_tag(x):\n",
    "    if x.split('_')[-1] in ['iPSCORE', 'HipSci']:\n",
    "        return x\n",
    "    else:\n",
    "        return x + '_Stitched'\n",
    "\n",
    "def get_stitch_cluster_info(combined, info):\n",
    "    def lambda_add_stitched_tag(x):\n",
    "        if x.split('_')[-1] in ['iPSCORE', 'HipSci', 'Stitched']:\n",
    "            return x\n",
    "        else:\n",
    "            return x + '_Stitched'\n",
    "    \n",
    "    \n",
    "    to_remove = []\n",
    "    for k in combined.keys():\n",
    "        to_remove += combined[k]\n",
    "    print('{} CNVs combined into {} CNVs.'.format(len(to_remove), len(combined)))\n",
    "    \n",
    "        \n",
    "    data = []\n",
    "    for i in combined.keys():\n",
    "        spl = i.split('_')\n",
    "        chrom = spl[1]\n",
    "        start,end = int(spl[2]), int(spl[3])\n",
    "        merged_calls = combined[i]\n",
    "        len_merge = len(merged_calls)\n",
    "        classes_merged = []\n",
    "        cluster_str = \",\".join(merged_calls)\n",
    "        \n",
    "        for z in merged_calls:\n",
    "            class_call = info.loc[z].cnv_class\n",
    "            classes_merged.append(class_call)\n",
    "     \n",
    "        data.append([chrom, start, end, i, len_merge, merged_calls, classes_merged, cluster_str])\n",
    "\n",
    "    merged = pd.DataFrame(data, columns=['chrom', 'start', 'end', 'stitched_cnv_site_ID','num_cnvs_merged', 'cnvs_merged', 'cnv_classes', 'stitch_cluster'])\n",
    "    \n",
    "    merged.sort_values(['chrom', 'start', 'end'], inplace=True)\n",
    "    \n",
    "    merged['merge_class']= merged.cnv_classes.apply(lambda x: characterize_cnv_classes(x))\n",
    "    merged['Length']= merged.end - merged.start\n",
    "    merged.index = merged.stitched_cnv_site_ID\n",
    "    \n",
    "    \n",
    "        \n",
    "    distances = []\n",
    "    for i in combined.keys():\n",
    "        list_sites = combined[i]\n",
    "        num_combined = len(list_sites)\n",
    "\n",
    "        for l,z in enumerate(list_sites[:-1]):\n",
    "            p1, p2 = list_sites[l],list_sites[l+1]\n",
    "            end_1 = int(p1.split('_')[3])\n",
    "            start_2 = int(p2.split('_')[2])\n",
    "\n",
    "            pair = p1 + ',' + p2\n",
    "\n",
    "            dist = start_2 - end_1\n",
    "            CNV_Types = str(info.loc[p1].cnv_class) + ',' + str(info.loc[p2].cnv_class)\n",
    "\n",
    "            distances.append([pair, i, CNV_Types, dist, num_combined])\n",
    "            \n",
    "\n",
    "    dist_frame = pd.DataFrame(distances, columns=['pair', 'stitched_cnv_site_ID' ,'cnv_types', 'distance', 'num_combined'])\n",
    "    mean_dist = dist_frame.groupby('stitched_cnv_site_ID').distance.mean().to_frame()\n",
    "    mean_dist['mean_distance_between']= mean_dist.distance\n",
    "    merged = merged.join(mean_dist['mean_distance_between'])\n",
    "    \n",
    "    \n",
    "    \n",
    "    cluster_ID_dict = {}\n",
    "    stitched_ID_dict = {}\n",
    "    count = 1\n",
    "    for x1, x2 in zip(merged.cnvs_merged.tolist(), merged['stitched_cnv_site_ID'].tolist()):\n",
    "        for ID in x1:\n",
    "            cluster_ID_dict[ID] = count\n",
    "            stitched_ID_dict[ID] = x2\n",
    "\n",
    "        cluster_ID_dict[x2] = count \n",
    "        count +=1\n",
    "\n",
    "    merged['stitch_cluster_ID'] = merged.stitched_cnv_site_ID.apply(lambda x: cluster_ID_dict[x])\n",
    "    \n",
    "    data = merged.apply(lambda x: lambda_stitch_site_in_cnvs_merged(x), axis = 1).tolist()\n",
    "    tdf = pd.DataFrame(data, columns=['stitch_site_already_gt', 'stitch_site_original_data', 'stitched_cnv'])\n",
    "    tdf.index = tdf.stitched_cnv\n",
    "    merged = merged.join(tdf)\n",
    "   \n",
    "    \n",
    "    merged['stitched_cnv_site_ID_mod'] = merged.stitched_cnv_site_ID.apply(lambda x: lambda_add_stitched_tag(x))\n",
    "    ind = merged[merged.stitch_site_already_gt == True].index.tolist()\n",
    "    merged.loc[ind, 'stitched_cnv_site_ID_mod']= merged.loc[ind, 'stitch_site_original_data']\n",
    "    \n",
    "    \n",
    "    cluster_ID_dict = {}\n",
    "    stitched_ID_dict_mod = {}\n",
    "    cluster_str_dict = {}\n",
    "    count = 1\n",
    "    for x1, x2, clust in zip(merged.cnvs_merged.tolist(), merged['stitched_cnv_site_ID_mod'].tolist(), merged.stitch_cluster.tolist()):\n",
    "        for ID in x1:\n",
    "            cluster_ID_dict[ID] = count\n",
    "            stitched_ID_dict_mod[ID] = x2\n",
    "            cluster_str_dict[ID] = clust\n",
    "\n",
    "\n",
    "    cluster_ID_dict[x2] = count \n",
    "    cluster_str_dict[x2] = clust\n",
    "    count +=1\n",
    "\n",
    "\n",
    "  \n",
    "    \n",
    "    # build the info data for all sites\n",
    "    tdf = pd.DataFrame([x.split('_') for x in combined.keys()], columns=['cnv', 'Chr', 'Start', 'End'],\n",
    "                   index=combined.keys()).drop('cnv', axis=1)\n",
    "   \n",
    "    tdf['ID']= tdf.index\n",
    "    tdf = tdf.join(merged[['stitch_cluster_ID', 'stitch_cluster', 'stitched_cnv_site_ID', 'stitched_cnv_site_ID_mod', 'stitch_site_original_data', 'stitch_site_already_gt']])\n",
    "    \n",
    "\n",
    "    tdf.index = tdf['stitched_cnv_site_ID_mod']\n",
    "    \n",
    "    tdf.ID = tdf.stitched_cnv_site_ID_mod\n",
    "    cols = ['Chr','Start', 'End', 'ID','stitch_breakpoint', \n",
    "                             'stitch_constituent', 'stitch_cluster_ID', 'stitch_cluster',\n",
    "            'stitched_cnv_site_ID', 'stitched_cnv_site_ID_mod', 'stitch_site_already_gt']\n",
    "    \n",
    "    tdf.Start = tdf.Start.astype(int)\n",
    "    tdf.End = tdf.End.astype(int)\n",
    "    # site generated by stitching?\n",
    "    tdf['stitch_breakpoint']=True\n",
    "    # is this a site that is a constituent of a stitching cluster?\n",
    "    tdf['stitch_constituent'] = False\n",
    "    \n",
    "\n",
    "\n",
    "    tdf = tdf[tdf.stitch_site_original_data == False]\n",
    "    tdf = tdf[cols]\n",
    "    \n",
    "    # what cluster of stitched site does this correspond to- stitched site ID also gets this numeric ID\n",
    "    \n",
    "    info['stitch_cluster_ID'] = info.ID.apply(lambda x: cluster_ID_dict.get(x, 0))\n",
    "    \n",
    "    # what stitch site is this site a constituent of, if any?\n",
    "    info['stitched_cnv_site_ID'] = info.ID.apply(lambda x: stitched_ID_dict.get(x, False))\n",
    "    info['stitched_cnv_site_ID_mod'] = info.ID.apply(lambda x: stitched_ID_dict_mod.get(x, False))\n",
    "    info['stitch_cluster'] = info.ID.apply(lambda x: cluster_str_dict.get(x, False))\n",
    "    \n",
    "    \n",
    "    \n",
    "    # mark sites in the original info that are stitch sites, but aren't new break points- (didn't need new genotyping)\n",
    "    # these are sites that are contained within one another mostly that didn't pass redundancy checking threshlods\n",
    "    \n",
    "    inds = list(set(merged[merged.stitch_site_original_data !=False].stitch_site_original_data.tolist()))\n",
    "    info['stitch_site_already_gt'] = False\n",
    "    info.loc[inds, 'stitch_site_already_gt' ] = True\n",
    "    \n",
    "    # mark constituent sites of stitching sites\n",
    "    info['stitch_constituent']= False\n",
    "    to_remove_mod = [i for i in to_remove if i not in inds]\n",
    "    info.loc[to_remove_mod, 'stitch_constituent'] = True\n",
    "    \n",
    "    \n",
    "    # mark these sites as stitch_breakpoint so we can still identify them, but allow them to keep their original IDs\n",
    "    # original IDs still carry info about the origin of the variant call (iPSCORE or HipSci discovery)\n",
    "    info['stitch_breakpoint'] = False\n",
    "    info.loc[inds, 'stitch_breakpoint' ] = True\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    # carry over original info col from the original sites that are unstitched\n",
    "    info_trunc = info[['Chr','Start', 'End', 'ID','stitch_breakpoint', \n",
    "                             'stitch_constituent', 'stitch_cluster_ID', 'stitch_cluster','stitched_cnv_site_ID', 'stitched_cnv_site_ID_mod', 'stitch_site_already_gt']]\n",
    "    \n",
    "    \n",
    "    gs_combined_info_unannotated = pd.concat([tdf, info_trunc])\n",
    "    gs_combined_info_unannotated = gs_combined_info_unannotated.sort_values(by=['Chr', 'Start', 'End'])\n",
    "    \n",
    "    \n",
    "    \n",
    "#     gs_combined_info_unannotated.ID =  gs_combined_info_unannotated.ID.apply(lambda x: lambda_add_stitched_tag(x))\n",
    "#     gs_combined_info_unannotated.index = gs_combined_info_unanno\n",
    "\n",
    "    \n",
    "    return merged, to_remove, dist_frame, gs_combined_info_unannotated, tdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_info_adj_sites(info, cns_t, sample_info):\n",
    "    \n",
    "    samples_discovery = sample_info[sample_info.CELL_TYPE != 'iPSC'].WGS_ID.tolist()\n",
    "    samples_ipscore_males = sample_info[(sample_info.STUDY == 'iPSCORE') & (sample_info.SEX == 'M')].WGS_ID.tolist()\n",
    "    samples_females = sample_info[(sample_info.CELL_TYPE != 'iPSC') & (sample_info.SEX=='F')].WGS_ID.tolist()\n",
    "    samples_males = sample_info[(sample_info.CELL_TYPE != 'iPSC') & (sample_info.SEX=='M')].WGS_ID.tolist()\n",
    "\n",
    "    \n",
    "    adj_sites = collect_data_adjacent_sites(info, cns_t, samples_discovery, subtract_lq=True).pipe(prep_adj_sites, thresh=30000)\n",
    "    passing_lq_correct = adj_sites[adj_sites.passing_criteria == True].index.tolist()\n",
    "\n",
    "    adj_sites_all_males = collect_data_adjacent_sites(info, cns_t, samples_males, cn_mode_col='cn_mode_male', subtract_lq=True).pipe(prep_adj_sites, thresh=30000)\n",
    "\n",
    "    adj_sites_all_females = collect_data_adjacent_sites(info, cns_t, samples_females, cn_mode_col='cn_mode_female', subtract_lq=True).pipe(prep_adj_sites, thresh=30000)\n",
    "\n",
    "    adj_sites_male_ipscore = collect_data_adjacent_sites(info, cns_t, samples_ipscore_males, cn_mode_col = 'cn_mode_male_ipscore_fb', subtract_lq=True).pipe(prep_adj_sites, thresh=30000)\n",
    "\n",
    "    x_chrom_consensus = gather_consensus_X(adj_sites_all_males, adj_sites_all_females, adj_sites, cns_t)\n",
    "\n",
    "    adj_sites['passing_criteria_adjusted'] = adj_sites.passing_criteria\n",
    "    inds_x = adj_sites[adj_sites.chrom=='X'].index.tolist()\n",
    "    inds_y = adj_sites[adj_sites.chrom=='Y'].index.tolist()\n",
    "    adj_sites_y_males = adj_sites_male_ipscore[adj_sites_male_ipscore.chrom=='Y'].copy()\n",
    "\n",
    "    adj_sites.loc[inds_x, 'passing_criteria_adjusted'] = False\n",
    "    adj_sites.loc[inds_y, 'passing_criteria_adjusted'] = False\n",
    "\n",
    "    adj_sites.loc[inds_x, 'passing_criteria_adjusted'] = x_chrom_consensus.passing_consensus\n",
    "    adj_sites.loc[inds_y, 'passing_criteria_adjusted'] = adj_sites_y_males.passing_criteria\n",
    "    \n",
    "    return adj_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/frazer01/home/djakubosky/software/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:25: RuntimeWarning: Mean of empty slice.\n"
     ]
    }
   ],
   "source": [
    "info = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/info_all_sites_rmdup_filt.pkl').pipe(prep_info)\n",
    "\n",
    "cns = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/cns_all.pkl').pipe(prep_cns, info)\n",
    "cns.drop('old_index', axis =1, inplace=True)\n",
    "\n",
    "sample_info = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/sample_info_combined/sample_info.pkl')\n",
    "\n",
    "samples_discovery = sample_info[sample_info.CELL_TYPE != 'iPSC'].WGS_ID.tolist()\n",
    "\n",
    "cns_t = cns.T.copy()\n",
    "cns_t = cns_t.loc[samples_discovery]\n",
    "\n",
    "adj_sites = get_info_adj_sites(info, cns_t, sample_info)\n",
    "\n",
    "to_combine  = compile_stitch_clusters(adj_sites, info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3913 CNVs combined into 1252 CNVs.\n"
     ]
    }
   ],
   "source": [
    "merged, to_remove, dist_frame, gs_combined_info_unannotated, t = get_stitch_cluster_info(to_combine, info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/frazer01/home/djakubosky/software/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:25: RuntimeWarning: Mean of empty slice.\n"
     ]
    }
   ],
   "source": [
    "info = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/gs_info_filt_rmdup.pkl').pipe(prep_info)\n",
    "\n",
    "cns = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/cns_all.pkl').pipe(prep_cns, info)\n",
    "cns.drop('old_index', axis =1, inplace=True)\n",
    "\n",
    "sample_info = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/sample_info_combined/sample_info.pkl')\n",
    "\n",
    "# samples_discovery = sample_info[sample_info.CELL_TYPE != 'iPSC'].WGS_ID.tolist()\n",
    "\n",
    "cns_t = cns.T.copy()\n",
    "cns_t = cns_t.loc[samples_discovery]\n",
    "\n",
    "adj_sites = get_info_adj_sites(info, cns_t, sample_info)\n",
    "\n",
    "to_combine  = compile_stitch_clusters(adj_sites, info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3913 CNVs combined into 1252 CNVs.\n"
     ]
    }
   ],
   "source": [
    "merged, to_remove, dist_frame, gs_combined_info_unannotated, t = get_stitch_cluster_info(to_combine, info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "out_dir = '/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stitch_site_info_v3 = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/stitch_site_info_v3.pkl')\n",
      "stitch_site_info_v3 = pd.read_csv('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/stitch_site_info_v3.tsv', sep='\\t')\n",
      "# all vars recorded: /frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_saved_nb_variables.py\n",
      "# pickled vars recorded:/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_pickled_nb_variables.py\n"
     ]
    }
   ],
   "source": [
    "CM.save_dataframe('stitch_site_info_v3', merged, out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gs_info_stitched_ua = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/gs_info_stitched_ua.pkl')\n",
      "gs_info_stitched_ua = pd.read_csv('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/gs_info_stitched_ua.tsv', sep='\\t')\n",
      "# all vars recorded: /frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_saved_nb_variables.py\n",
      "# pickled vars recorded:/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_pickled_nb_variables.py\n"
     ]
    }
   ],
   "source": [
    "CM.save_dataframe('gs_info_stitched_ua', gs_combined_info_unannotated, out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adj_site_info = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/adj_site_info.pkl')\n",
      "adj_site_info = pd.read_csv('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/adj_site_info.tsv', sep='\\t')\n",
      "# all vars recorded: /frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_saved_nb_variables.py\n",
      "# pickled vars recorded:/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_pickled_nb_variables.py\n"
     ]
    }
   ],
   "source": [
    "CM.save_dataframe('adj_site_info', adj_sites, out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gs_info_stitched_ua = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/gs_info_stitched_ua.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3913 CNVs combined into 1252 CNVs.\n"
     ]
    }
   ],
   "source": [
    "merged, to_remove, dist_frame, gs_combined_info_unannotated, t = get_stitch_cluster_info(to_combine, info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merged, to_remove, dist_frame, gs_combined_info_unannotated, t = get_stitch_cluster_info(to_combine, info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    922\n",
       "True     397\n",
       "Name: stitch_site_already_gt, dtype: int64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# stitch_site_info_v2.stitch_site_already_gt.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    897\n",
       "True     355\n",
       "Name: stitch_site_already_gt, dtype: int64"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.stitch_site_already_gt.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stitch_site_info_v2 = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/stitch_site_info_v2.pkl')\n",
      "stitch_site_info_v2 = pd.read_csv('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/stitch_site_info_v2.tsv', sep='\\t')\n",
      "# all vars recorded: /frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_saved_nb_variables.py\n",
      "# pickled vars recorded:/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_pickled_nb_variables.py\n"
     ]
    }
   ],
   "source": [
    "CM.save_dataframe('stitch_site_info_v2', merged, out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gs_info_stitched_ua = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/gs_info_stitched_ua.pkl')\n",
      "gs_info_stitched_ua = pd.read_csv('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/gs_info_stitched_ua.tsv', sep='\\t')\n",
      "# all vars recorded: /frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_saved_nb_variables.py\n",
      "# pickled vars recorded:/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_pickled_nb_variables.py\n"
     ]
    }
   ],
   "source": [
    "CM.save_dataframe('gs_info_stitched_ua', gs_combined_info_unannotated, out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "adj_site_info = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/adj_site_info.pkl')\n",
      "adj_site_info = pd.read_csv('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/adj_site_info.tsv', sep='\\t')\n",
      "# all vars recorded: /frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_saved_nb_variables.py\n",
      "# pickled vars recorded:/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_pickled_nb_variables.py\n"
     ]
    }
   ],
   "source": [
    "CM.save_dataframe('adj_site_info', adj_sites, out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_site_info = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V3/i2QTL_combined/adj_site_info.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID1</th>\n",
       "      <th>ID2</th>\n",
       "      <th>corr_coef</th>\n",
       "      <th>num_diff</th>\n",
       "      <th>num_non_mode</th>\n",
       "      <th>percent_non_mode_diff</th>\n",
       "      <th>samps_to_compare_nmode</th>\n",
       "      <th>samps_to_compare_corr</th>\n",
       "      <th>samps_to_exclude</th>\n",
       "      <th>num_pass</th>\n",
       "      <th>exact_cn_match</th>\n",
       "      <th>mean_cn_diff_all</th>\n",
       "      <th>mean_cn_diff_nmode</th>\n",
       "      <th>alleles1</th>\n",
       "      <th>alleles2</th>\n",
       "      <th>num_alleles1</th>\n",
       "      <th>num_alleles2</th>\n",
       "      <th>allele_dist1</th>\n",
       "      <th>allele_dist2</th>\n",
       "      <th>pair</th>\n",
       "      <th>mode_cn1</th>\n",
       "      <th>mode_cn2</th>\n",
       "      <th>cnv_class1</th>\n",
       "      <th>cnv_class2</th>\n",
       "      <th>distance_between</th>\n",
       "      <th>distance_between_mod</th>\n",
       "      <th>cat_pair</th>\n",
       "      <th>num_diff1</th>\n",
       "      <th>num_diff2</th>\n",
       "      <th>diff_uuids1</th>\n",
       "      <th>diff_uuids2</th>\n",
       "      <th>chrom</th>\n",
       "      <th>matching_mode</th>\n",
       "      <th>log_dist</th>\n",
       "      <th>passing_criteria</th>\n",
       "      <th>passing_criteria_adjusted</th>\n",
       "      <th>order</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cat_pair</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CNV_11_114477669_114482169_HipSci-CNV_11_114478169_114481969_iPSCORE</th>\n",
       "      <td>CNV_11_114477669_114482169_HipSci</td>\n",
       "      <td>CNV_11_114478169_114481969_iPSCORE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[SAMEA2632917, SAMEA2555011, 0f518720-9d4b-498...</td>\n",
       "      <td>[fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA27...</td>\n",
       "      <td>[]</td>\n",
       "      <td>478</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{1, 2, 3}</td>\n",
       "      <td>{1, 2, 3}</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>{1: 7, 2: 468, 3: 3}</td>\n",
       "      <td>{1: 7, 2: 468, 3: 3}</td>\n",
       "      <td>[CNV_11_114477669_114482169_HipSci, CNV_11_114...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>-4000</td>\n",
       "      <td>0</td>\n",
       "      <td>CNV_11_114477669_114482169_HipSci-CNV_11_11447...</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>[ae757c28-7757-4a5e-9016-9d89dbf56499, 0f51872...</td>\n",
       "      <td>[ae757c28-7757-4a5e-9016-9d89dbf56499, 0f51872...</td>\n",
       "      <td>11</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2286</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                  ID1  \\\n",
       "cat_pair                                                                                \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  CNV_11_114477669_114482169_HipSci   \n",
       "\n",
       "                                                                                   ID2  \\\n",
       "cat_pair                                                                                 \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  CNV_11_114478169_114481969_iPSCORE   \n",
       "\n",
       "                                                    corr_coef  num_diff  \\\n",
       "cat_pair                                                                  \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...        1.0         0   \n",
       "\n",
       "                                                    num_non_mode  \\\n",
       "cat_pair                                                           \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...            10   \n",
       "\n",
       "                                                    percent_non_mode_diff  \\\n",
       "cat_pair                                                                    \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...                    0.0   \n",
       "\n",
       "                                                                               samps_to_compare_nmode  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  [SAMEA2632917, SAMEA2555011, 0f518720-9d4b-498...   \n",
       "\n",
       "                                                                                samps_to_compare_corr  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  [fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA27...   \n",
       "\n",
       "                                                   samps_to_exclude  num_pass  \\\n",
       "cat_pair                                                                        \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...               []       478   \n",
       "\n",
       "                                                    exact_cn_match  \\\n",
       "cat_pair                                                             \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...            True   \n",
       "\n",
       "                                                    mean_cn_diff_all  \\\n",
       "cat_pair                                                               \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...               0.0   \n",
       "\n",
       "                                                    mean_cn_diff_nmode  \\\n",
       "cat_pair                                                                 \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...                 0.0   \n",
       "\n",
       "                                                     alleles1   alleles2  \\\n",
       "cat_pair                                                                   \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  {1, 2, 3}  {1, 2, 3}   \n",
       "\n",
       "                                                    num_alleles1  \\\n",
       "cat_pair                                                           \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...             3   \n",
       "\n",
       "                                                    num_alleles2  \\\n",
       "cat_pair                                                           \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...             3   \n",
       "\n",
       "                                                            allele_dist1  \\\n",
       "cat_pair                                                                   \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  {1: 7, 2: 468, 3: 3}   \n",
       "\n",
       "                                                            allele_dist2  \\\n",
       "cat_pair                                                                   \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  {1: 7, 2: 468, 3: 3}   \n",
       "\n",
       "                                                                                                 pair  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  [CNV_11_114477669_114482169_HipSci, CNV_11_114...   \n",
       "\n",
       "                                                    mode_cn1  mode_cn2  \\\n",
       "cat_pair                                                                 \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...         2         2   \n",
       "\n",
       "                                                   cnv_class1 cnv_class2  \\\n",
       "cat_pair                                                                   \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...       mCNV       mCNV   \n",
       "\n",
       "                                                    distance_between  \\\n",
       "cat_pair                                                               \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...             -4000   \n",
       "\n",
       "                                                    distance_between_mod  \\\n",
       "cat_pair                                                                   \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...                     0   \n",
       "\n",
       "                                                                                             cat_pair  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  CNV_11_114477669_114482169_HipSci-CNV_11_11447...   \n",
       "\n",
       "                                                    num_diff1  num_diff2  \\\n",
       "cat_pair                                                                   \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...         10         10   \n",
       "\n",
       "                                                                                          diff_uuids1  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  [ae757c28-7757-4a5e-9016-9d89dbf56499, 0f51872...   \n",
       "\n",
       "                                                                                          diff_uuids2  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...  [ae757c28-7757-4a5e-9016-9d89dbf56499, 0f51872...   \n",
       "\n",
       "                                                   chrom  matching_mode  \\\n",
       "cat_pair                                                                  \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...    11           True   \n",
       "\n",
       "                                                    log_dist  \\\n",
       "cat_pair                                                       \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...       0.0   \n",
       "\n",
       "                                                    passing_criteria  \\\n",
       "cat_pair                                                               \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...              True   \n",
       "\n",
       "                                                    passing_criteria_adjusted  \\\n",
       "cat_pair                                                                        \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...                       True   \n",
       "\n",
       "                                                    order  \n",
       "cat_pair                                                   \n",
       "CNV_11_114477669_114482169_HipSci-CNV_11_114478...   2286  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adj_site_info[adj_site_info.ID1 == 'CNV_11_114477669_114482169_HipSci']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID1</th>\n",
       "      <th>ID2</th>\n",
       "      <th>corr_coef</th>\n",
       "      <th>num_diff</th>\n",
       "      <th>num_non_mode</th>\n",
       "      <th>percent_non_mode_diff</th>\n",
       "      <th>samps_to_compare_nmode</th>\n",
       "      <th>samps_to_compare_corr</th>\n",
       "      <th>samps_to_exclude</th>\n",
       "      <th>num_pass</th>\n",
       "      <th>exact_cn_match</th>\n",
       "      <th>mean_cn_diff_all</th>\n",
       "      <th>mean_cn_diff_nmode</th>\n",
       "      <th>alleles1</th>\n",
       "      <th>alleles2</th>\n",
       "      <th>num_alleles1</th>\n",
       "      <th>num_alleles2</th>\n",
       "      <th>allele_dist1</th>\n",
       "      <th>allele_dist2</th>\n",
       "      <th>pair</th>\n",
       "      <th>mode_cn1</th>\n",
       "      <th>mode_cn2</th>\n",
       "      <th>cnv_class1</th>\n",
       "      <th>cnv_class2</th>\n",
       "      <th>distance_between</th>\n",
       "      <th>distance_between_mod</th>\n",
       "      <th>cat_pair</th>\n",
       "      <th>num_diff1</th>\n",
       "      <th>num_diff2</th>\n",
       "      <th>diff_uuids1</th>\n",
       "      <th>diff_uuids2</th>\n",
       "      <th>chrom</th>\n",
       "      <th>matching_mode</th>\n",
       "      <th>log_dist</th>\n",
       "      <th>passing_criteria</th>\n",
       "      <th>passing_criteria_adjusted</th>\n",
       "      <th>order</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cat_pair</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPSCORE</th>\n",
       "      <td>CNV_1_55175_56686_iPSCORE</td>\n",
       "      <td>CNV_1_85391_86406_iPSCORE</td>\n",
       "      <td>0.846991</td>\n",
       "      <td>107</td>\n",
       "      <td>295</td>\n",
       "      <td>0.362712</td>\n",
       "      <td>[fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA32...</td>\n",
       "      <td>[fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA27...</td>\n",
       "      <td>[ee146263-82f5-4ed7-bb49-4300837acc85, d941dce...</td>\n",
       "      <td>438</td>\n",
       "      <td>False</td>\n",
       "      <td>0.273973</td>\n",
       "      <td>0.406780</td>\n",
       "      <td>{0, 1, 2, 3, 4, 5}</td>\n",
       "      <td>{0, 1, 2, 3, 4, 5}</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>{0: 116, 1: 194, 2: 97, 3: 25, 4: 4, 5: 2}</td>\n",
       "      <td>{0: 102, 1: 167, 2: 116, 3: 41, 4: 11, 5: 1}</td>\n",
       "      <td>[CNV_1_55175_56686_iPSCORE, CNV_1_85391_86406_...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>28705</td>\n",
       "      <td>28705</td>\n",
       "      <td>CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iP...</td>\n",
       "      <td>259</td>\n",
       "      <td>282</td>\n",
       "      <td>[61d77efd-eec4-421b-a717-4486512668da, 8316c9a...</td>\n",
       "      <td>[41b56a79-65a4-44d0-8f3e-86c7d6621fcb, c7a966f...</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>4.457973</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPSCORE</th>\n",
       "      <td>CNV_1_85391_86406_iPSCORE</td>\n",
       "      <td>CNV_1_86406_91299_iPSCORE</td>\n",
       "      <td>0.906808</td>\n",
       "      <td>84</td>\n",
       "      <td>295</td>\n",
       "      <td>0.284746</td>\n",
       "      <td>[fea585f4-3cae-48a1-8e2c-e89d88930b99, 3ba3d4d...</td>\n",
       "      <td>[fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA27...</td>\n",
       "      <td>[ee146263-82f5-4ed7-bb49-4300837acc85, d941dce...</td>\n",
       "      <td>450</td>\n",
       "      <td>False</td>\n",
       "      <td>0.202222</td>\n",
       "      <td>0.308475</td>\n",
       "      <td>{0, 1, 2, 3, 4, 5}</td>\n",
       "      <td>{0, 1, 2, 3, 4}</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>{0: 102, 1: 169, 2: 119, 3: 44, 4: 15, 5: 1}</td>\n",
       "      <td>{0: 116, 1: 196, 2: 99, 3: 33, 4: 6}</td>\n",
       "      <td>[CNV_1_85391_86406_iPSCORE, CNV_1_86406_91299_...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iP...</td>\n",
       "      <td>282</td>\n",
       "      <td>272</td>\n",
       "      <td>[41b56a79-65a4-44d0-8f3e-86c7d6621fcb, c7a966f...</td>\n",
       "      <td>[32ee31cb-156d-4a95-a708-6271f7e2cbab, 8316c9a...</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_iPSCORE</th>\n",
       "      <td>CNV_1_86406_91299_iPSCORE</td>\n",
       "      <td>CNV_1_564439_567804_iPSCORE</td>\n",
       "      <td>0.049197</td>\n",
       "      <td>238</td>\n",
       "      <td>345</td>\n",
       "      <td>0.689855</td>\n",
       "      <td>[fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA27...</td>\n",
       "      <td>[fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA27...</td>\n",
       "      <td>[4a1becf2-f540-4a14-99a2-41aa3205bd81, SAMEA31...</td>\n",
       "      <td>435</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372414</td>\n",
       "      <td>1.469565</td>\n",
       "      <td>{0, 1, 2, 3, 4}</td>\n",
       "      <td>{0, 1, 2, 3, 4, 5, 6, 7, 9, 12, 13, 15, 20}</td>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "      <td>{0: 103, 1: 187, 2: 101, 3: 36, 4: 8}</td>\n",
       "      <td>{0: 12, 1: 101, 2: 219, 3: 70, 4: 14, 5: 9, 6:...</td>\n",
       "      <td>[CNV_1_86406_91299_iPSCORE, CNV_1_564439_56780...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>473140</td>\n",
       "      <td>473140</td>\n",
       "      <td>CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_...</td>\n",
       "      <td>272</td>\n",
       "      <td>216</td>\n",
       "      <td>[32ee31cb-156d-4a95-a708-6271f7e2cbab, 8316c9a...</td>\n",
       "      <td>[5a150d58-df23-4b02-b708-c6594b180d14, 41b56a7...</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>5.674991</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497_iPSCORE</th>\n",
       "      <td>CNV_1_564439_567804_iPSCORE</td>\n",
       "      <td>CNV_1_565396_567497_iPSCORE</td>\n",
       "      <td>0.699246</td>\n",
       "      <td>37</td>\n",
       "      <td>210</td>\n",
       "      <td>0.176190</td>\n",
       "      <td>[3ba3d4d5-9589-4f74-aa4c-28a96f8c222b, SAMEA32...</td>\n",
       "      <td>[SAMEA2784668, 2914ab72-1d02-470a-ba53-7868982...</td>\n",
       "      <td>[fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA23...</td>\n",
       "      <td>420</td>\n",
       "      <td>False</td>\n",
       "      <td>0.245238</td>\n",
       "      <td>0.490476</td>\n",
       "      <td>{0, 1, 2, 3, 4, 5, 6, 7, 12, 13, 15, 20}</td>\n",
       "      <td>{0, 1, 2, 3, 4, 5, 8, 14}</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>{0: 12, 1: 101, 2: 216, 3: 65, 4: 11, 5: 7, 6:...</td>\n",
       "      <td>{0: 13, 1: 106, 2: 217, 3: 71, 4: 10, 5: 1, 8:...</td>\n",
       "      <td>[CNV_1_564439_567804_iPSCORE, CNV_1_565396_567...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>-2408</td>\n",
       "      <td>0</td>\n",
       "      <td>CNV_1_564439_567804_iPSCORE-CNV_1_565396_56749...</td>\n",
       "      <td>216</td>\n",
       "      <td>233</td>\n",
       "      <td>[5a150d58-df23-4b02-b708-c6594b180d14, 41b56a7...</td>\n",
       "      <td>[5a150d58-df23-4b02-b708-c6594b180d14, dd70912...</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904_HipSci</th>\n",
       "      <td>CNV_1_565396_567497_iPSCORE</td>\n",
       "      <td>CNV_1_565396_567904_HipSci</td>\n",
       "      <td>0.558123</td>\n",
       "      <td>8</td>\n",
       "      <td>222</td>\n",
       "      <td>0.036036</td>\n",
       "      <td>[SAMEA3257661, SAMEA2438671, 4df98c9a-93f7-473...</td>\n",
       "      <td>[SAMEA2784668, 2914ab72-1d02-470a-ba53-7868982...</td>\n",
       "      <td>[fea585f4-3cae-48a1-8e2c-e89d88930b99, 3ba3d4d...</td>\n",
       "      <td>440</td>\n",
       "      <td>False</td>\n",
       "      <td>0.143182</td>\n",
       "      <td>0.283784</td>\n",
       "      <td>{0, 1, 2, 3, 4, 8, 14}</td>\n",
       "      <td>{0, 1, 2, 3, 4, 5, 7, 13, 16, 18}</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>{0: 15, 1: 111, 2: 221, 3: 78, 4: 13, 8: 1, 14...</td>\n",
       "      <td>{0: 14, 1: 109, 2: 218, 3: 78, 4: 13, 5: 1, 7:...</td>\n",
       "      <td>[CNV_1_565396_567497_iPSCORE, CNV_1_565396_567...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>mCNV</td>\n",
       "      <td>-2101</td>\n",
       "      <td>0</td>\n",
       "      <td>CNV_1_565396_567497_iPSCORE-CNV_1_565396_56790...</td>\n",
       "      <td>233</td>\n",
       "      <td>227</td>\n",
       "      <td>[5a150d58-df23-4b02-b708-c6594b180d14, dd70912...</td>\n",
       "      <td>[5a150d58-df23-4b02-b708-c6594b180d14, dd70912...</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                            ID1  \\\n",
       "cat_pair                                                                          \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...    CNV_1_55175_56686_iPSCORE   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...    CNV_1_85391_86406_iPSCORE   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...    CNV_1_86406_91299_iPSCORE   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  CNV_1_564439_567804_iPSCORE   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  CNV_1_565396_567497_iPSCORE   \n",
       "\n",
       "                                                                            ID2  \\\n",
       "cat_pair                                                                          \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...    CNV_1_85391_86406_iPSCORE   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...    CNV_1_86406_91299_iPSCORE   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  CNV_1_564439_567804_iPSCORE   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  CNV_1_565396_567497_iPSCORE   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...   CNV_1_565396_567904_HipSci   \n",
       "\n",
       "                                                    corr_coef  num_diff  \\\n",
       "cat_pair                                                                  \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...   0.846991       107   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...   0.906808        84   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...   0.049197       238   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...   0.699246        37   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...   0.558123         8   \n",
       "\n",
       "                                                    num_non_mode  \\\n",
       "cat_pair                                                           \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...           295   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...           295   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...           345   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...           210   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...           222   \n",
       "\n",
       "                                                    percent_non_mode_diff  \\\n",
       "cat_pair                                                                    \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...               0.362712   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...               0.284746   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...               0.689855   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...               0.176190   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...               0.036036   \n",
       "\n",
       "                                                                               samps_to_compare_nmode  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...  [fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA32...   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...  [fea585f4-3cae-48a1-8e2c-e89d88930b99, 3ba3d4d...   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  [fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA27...   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  [3ba3d4d5-9589-4f74-aa4c-28a96f8c222b, SAMEA32...   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  [SAMEA3257661, SAMEA2438671, 4df98c9a-93f7-473...   \n",
       "\n",
       "                                                                                samps_to_compare_corr  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...  [fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA27...   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...  [fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA27...   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  [fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA27...   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  [SAMEA2784668, 2914ab72-1d02-470a-ba53-7868982...   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  [SAMEA2784668, 2914ab72-1d02-470a-ba53-7868982...   \n",
       "\n",
       "                                                                                     samps_to_exclude  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...  [ee146263-82f5-4ed7-bb49-4300837acc85, d941dce...   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...  [ee146263-82f5-4ed7-bb49-4300837acc85, d941dce...   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  [4a1becf2-f540-4a14-99a2-41aa3205bd81, SAMEA31...   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  [fea585f4-3cae-48a1-8e2c-e89d88930b99, SAMEA23...   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  [fea585f4-3cae-48a1-8e2c-e89d88930b99, 3ba3d4d...   \n",
       "\n",
       "                                                    num_pass  exact_cn_match  \\\n",
       "cat_pair                                                                       \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...       438           False   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...       450           False   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...       435           False   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...       420           False   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...       440           False   \n",
       "\n",
       "                                                    mean_cn_diff_all  \\\n",
       "cat_pair                                                               \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...          0.273973   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...          0.202222   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...          1.372414   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...          0.245238   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...          0.143182   \n",
       "\n",
       "                                                    mean_cn_diff_nmode  \\\n",
       "cat_pair                                                                 \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...            0.406780   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...            0.308475   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...            1.469565   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...            0.490476   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...            0.283784   \n",
       "\n",
       "                                                                                    alleles1  \\\n",
       "cat_pair                                                                                       \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...                        {0, 1, 2, 3, 4, 5}   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...                        {0, 1, 2, 3, 4, 5}   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...                           {0, 1, 2, 3, 4}   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  {0, 1, 2, 3, 4, 5, 6, 7, 12, 13, 15, 20}   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...                    {0, 1, 2, 3, 4, 8, 14}   \n",
       "\n",
       "                                                                                       alleles2  \\\n",
       "cat_pair                                                                                          \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...                           {0, 1, 2, 3, 4, 5}   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...                              {0, 1, 2, 3, 4}   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  {0, 1, 2, 3, 4, 5, 6, 7, 9, 12, 13, 15, 20}   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...                    {0, 1, 2, 3, 4, 5, 8, 14}   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...            {0, 1, 2, 3, 4, 5, 7, 13, 16, 18}   \n",
       "\n",
       "                                                    num_alleles1  \\\n",
       "cat_pair                                                           \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...             6   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...             6   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...             5   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...            12   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...             7   \n",
       "\n",
       "                                                    num_alleles2  \\\n",
       "cat_pair                                                           \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...             6   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...             5   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...            13   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...             8   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...            10   \n",
       "\n",
       "                                                                                         allele_dist1  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...         {0: 116, 1: 194, 2: 97, 3: 25, 4: 4, 5: 2}   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...       {0: 102, 1: 169, 2: 119, 3: 44, 4: 15, 5: 1}   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...              {0: 103, 1: 187, 2: 101, 3: 36, 4: 8}   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  {0: 12, 1: 101, 2: 216, 3: 65, 4: 11, 5: 7, 6:...   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  {0: 15, 1: 111, 2: 221, 3: 78, 4: 13, 8: 1, 14...   \n",
       "\n",
       "                                                                                         allele_dist2  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...       {0: 102, 1: 167, 2: 116, 3: 41, 4: 11, 5: 1}   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...               {0: 116, 1: 196, 2: 99, 3: 33, 4: 6}   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  {0: 12, 1: 101, 2: 219, 3: 70, 4: 14, 5: 9, 6:...   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  {0: 13, 1: 106, 2: 217, 3: 71, 4: 10, 5: 1, 8:...   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  {0: 14, 1: 109, 2: 218, 3: 78, 4: 13, 5: 1, 7:...   \n",
       "\n",
       "                                                                                                 pair  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...  [CNV_1_55175_56686_iPSCORE, CNV_1_85391_86406_...   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...  [CNV_1_85391_86406_iPSCORE, CNV_1_86406_91299_...   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  [CNV_1_86406_91299_iPSCORE, CNV_1_564439_56780...   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  [CNV_1_564439_567804_iPSCORE, CNV_1_565396_567...   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  [CNV_1_565396_567497_iPSCORE, CNV_1_565396_567...   \n",
       "\n",
       "                                                    mode_cn1  mode_cn2  \\\n",
       "cat_pair                                                                 \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...         1         1   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...         1         1   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...         1         2   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...         2         2   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...         2         2   \n",
       "\n",
       "                                                   cnv_class1 cnv_class2  \\\n",
       "cat_pair                                                                   \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...       mCNV       mCNV   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...       mCNV       mCNV   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...       mCNV       mCNV   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...       mCNV       mCNV   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...       mCNV       mCNV   \n",
       "\n",
       "                                                    distance_between  \\\n",
       "cat_pair                                                               \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...             28705   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...                 0   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...            473140   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...             -2408   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...             -2101   \n",
       "\n",
       "                                                    distance_between_mod  \\\n",
       "cat_pair                                                                   \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...                 28705   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...                     0   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...                473140   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...                     0   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...                     0   \n",
       "\n",
       "                                                                                             cat_pair  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...  CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iP...   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...  CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iP...   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_...   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  CNV_1_564439_567804_iPSCORE-CNV_1_565396_56749...   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  CNV_1_565396_567497_iPSCORE-CNV_1_565396_56790...   \n",
       "\n",
       "                                                    num_diff1  num_diff2  \\\n",
       "cat_pair                                                                   \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...        259        282   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...        282        272   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...        272        216   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...        216        233   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...        233        227   \n",
       "\n",
       "                                                                                          diff_uuids1  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...  [61d77efd-eec4-421b-a717-4486512668da, 8316c9a...   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...  [41b56a79-65a4-44d0-8f3e-86c7d6621fcb, c7a966f...   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  [32ee31cb-156d-4a95-a708-6271f7e2cbab, 8316c9a...   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  [5a150d58-df23-4b02-b708-c6594b180d14, 41b56a7...   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  [5a150d58-df23-4b02-b708-c6594b180d14, dd70912...   \n",
       "\n",
       "                                                                                          diff_uuids2  \\\n",
       "cat_pair                                                                                                \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...  [41b56a79-65a4-44d0-8f3e-86c7d6621fcb, c7a966f...   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...  [32ee31cb-156d-4a95-a708-6271f7e2cbab, 8316c9a...   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  [5a150d58-df23-4b02-b708-c6594b180d14, 41b56a7...   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  [5a150d58-df23-4b02-b708-c6594b180d14, dd70912...   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  [5a150d58-df23-4b02-b708-c6594b180d14, dd70912...   \n",
       "\n",
       "                                                   chrom  matching_mode  \\\n",
       "cat_pair                                                                  \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...     1           True   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...     1           True   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...     1          False   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...     1           True   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...     1           True   \n",
       "\n",
       "                                                    log_dist  \\\n",
       "cat_pair                                                       \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...  4.457973   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...  0.000000   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...  5.674991   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...  0.000000   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...  0.000000   \n",
       "\n",
       "                                                    passing_criteria  \\\n",
       "cat_pair                                                               \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...             False   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...             False   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...             False   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...             False   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...             False   \n",
       "\n",
       "                                                    passing_criteria_adjusted  \\\n",
       "cat_pair                                                                        \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...                      False   \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...                      False   \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...                      False   \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...                      False   \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...                      False   \n",
       "\n",
       "                                                    order  \n",
       "cat_pair                                                   \n",
       "CNV_1_55175_56686_iPSCORE-CNV_1_85391_86406_iPS...      0  \n",
       "CNV_1_85391_86406_iPSCORE-CNV_1_86406_91299_iPS...      1  \n",
       "CNV_1_86406_91299_iPSCORE-CNV_1_564439_567804_i...      2  \n",
       "CNV_1_564439_567804_iPSCORE-CNV_1_565396_567497...      3  \n",
       "CNV_1_565396_567497_iPSCORE-CNV_1_565396_567904...      4  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adj_site_info['']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "adj_site_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "info = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_combine_ipscore_hipsci/i2QTL_final/info_all_sites_rmdup_filt.pkl').pipe(prep_info)\n",
    "cns = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V3/i2QTL_combined/cns_all.pkl').pipe(prep_cns, info)\n",
    "cns.drop('old_index', axis =1, inplace=True)\n",
    "\n",
    "sample_info = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/sample_info_combined/sample_info.pkl')\n",
    "\n",
    "\n",
    "cns_t = cns.T.copy()\n",
    "cns_t = cns_t.loc[samples_discovery]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/frazer01/home/djakubosky/software/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:25: RuntimeWarning: Mean of empty slice.\n"
     ]
    }
   ],
   "source": [
    "adj_sites = get_info_adj_sites(info, cns_t, sample_info)\n",
    "\n",
    "to_combine  = compile_stitch_clusters(adj_sites, info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3913 CNVs combined into 1252 CNVs.\n"
     ]
    }
   ],
   "source": [
    "merged, to_remove, dist_frame, gs_combined_info_unannotated, t = get_stitch_cluster_info(to_combine, info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = '/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    897\n",
       "True     355\n",
       "Name: stitch_site_already_gt, dtype: int64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.stitch_site_already_gt.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stitch_site_info_v2 = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/stitch_site_info_v2.pkl')\n",
      "stitch_site_info_v2 = pd.read_csv('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/stitch_site_info_v2.tsv', sep='\\t')\n",
      "# all vars recorded: /frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_saved_nb_variables.py\n",
      "# pickled vars recorded:/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_pickled_nb_variables.py\n",
      "gs_info_stitched_ua = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/gs_info_stitched_ua.pkl')\n",
      "gs_info_stitched_ua = pd.read_csv('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/gs_info_stitched_ua.tsv', sep='\\t')\n",
      "# all vars recorded: /frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_saved_nb_variables.py\n",
      "# pickled vars recorded:/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_pickled_nb_variables.py\n",
      "adj_site_info = pd.read_pickle('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/adj_site_info.pkl')\n",
      "adj_site_info = pd.read_csv('/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/adj_site_info.tsv', sep='\\t')\n",
      "# all vars recorded: /frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_saved_nb_variables.py\n",
      "# pickled vars recorded:/frazer01/projects/hipsci/analysis/i2QTL-sv-analysis/private_output/gs_processing_V4/i2QTL_combined/load_pickled_nb_variables.py\n"
     ]
    }
   ],
   "source": [
    "CM.save_dataframe('stitch_site_info_v2', merged, out_dir)\n",
    "CM.save_dataframe('gs_info_stitched_ua', gs_combined_info_unannotated, out_dir)\n",
    "CM.save_dataframe('adj_site_info', adj_sites, out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Deprecated/Development"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix info data for the stitched cnv sites that are already present in the data set\n",
    "\n",
    "alread_gt_ind = merged[merged.stitch_site_already_gt == True].stitched_cnv_site_ID_mod.tolist()\n",
    "\n",
    "gs_combined_info_unannotated['stitch_site_already_gt'] = False\n",
    "gs_combined_info_unannotated.loc[alread_gt_ind] = True\n",
    "gs_combined_info_unannotated['stitched_cnv_site_ID_mod'] = False\n",
    "\n",
    "\n",
    "tdf = merged[merged.stitch_site_already_gt == True].copy()\n",
    "# tdf.index = tdf.stitch_site_already_gt\n",
    "\n",
    "tdf.index = tdf.stitch_site_original_data\n",
    "\n",
    "tdf['stitch_breakpoint'] = True\n",
    "tdf['stitch_constituent'] = False\n",
    "tdf['stitch_site_already_gt'] = False\n",
    "\n",
    "inds_original_in_stitch = merged[merged.stitch_site_already_gt == True].stitch_site_original_data.tolist()\n",
    "\n",
    "cols = 'cluster_ID\tcluster_str\tstitch_breakpoint\tstitch_constituent\tstitched_cnv_site_ID\tstitch_site_already_gt stitched_cnv_site_ID_mod'.split()\n",
    "\n",
    "test = tdf.loc[inds_original_in_stitch][cols].values\n",
    "\n",
    "gs_combined_info_unannotated.loc[inds_original_in_stitch, cols] = test\n",
    "\n",
    "gs_combined_info_unannotated.stitch_breakpoint.value_counts()\n",
    "\n",
    "gs_combined_info_unannotated = gs_combined_info_unannotated[gs_combined_info_unannotated.stitch_site_already_gt == False]\n",
    "\n",
    "\n",
    "gs_combined_info_unannotated.loc[inds_original_in_stitch, 'stitch_site_already_gt'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# def gather_consensus_X(adj_sites_all_males, adj_sites_all_females, adj_sites, cns_t):\n",
    "#     inds = adj_sites_all_males.index.tolist()\n",
    "#     assert (inds == adj_sites_all_females.index.tolist())\n",
    "    \n",
    "#     num_non_mode_males = adj_sites_all_males.num_non_mode.tolist()\n",
    "#     num_non_mode_females = adj_sites_all_females.num_non_mode.tolist()\n",
    "    \n",
    "#     ndiff_male = adj_sites_all_males.num_diff.tolist()\n",
    "#     ndiff_female = adj_sites_all_females.num_diff.tolist()\n",
    "    \n",
    "#     passing_b_males = adj_sites_all_males.passing_criteria.tolist()\n",
    "#     passing_b_females = adj_sites_all_females.passing_criteria.tolist()\n",
    "    \n",
    "#     samples_nm_males = adj_sites_all_males.samps_to_compare_corr.tolist()\n",
    "#     samples_nm_females = adj_sites_all_females.samps_to_compare_corr.tolist()\n",
    "#     union_samples = [list(set(i1 + i2)) for i1, i2 in zip(samples_nm_males, samples_nm_females)]\n",
    "   \n",
    "#     corr_males = adj_sites_all_males.corr_coef.tolist()\n",
    "#     corr_females = adj_sites_all_females.corr_coef.tolist()\n",
    "#     dist = adj_sites_all_males.distance_between_mod.tolist()\n",
    "    \n",
    "#     out = []\n",
    "#     for ind, nm_male, nm_female, pass_b_male, pass_b_female, cm, cf, ndm, ndf, d, us in zip(inds,\n",
    "#                                                                                             num_non_mode_males,\n",
    "#                                                                            num_non_mode_females,\n",
    "#                                                                            passing_b_males, passing_b_females, \n",
    "#                                                                            corr_males, corr_females,\n",
    "#                                                                            ndiff_male, ndiff_female, \n",
    "#                                                                            dist, union_samples):\n",
    "        \n",
    "        \n",
    "#         ind1, ind2 = ind.split('-')\n",
    "#         chrom = ind.split('_')[1]\n",
    "        \n",
    "#         if chrom == 'X':\n",
    "\n",
    "#             if [nm_male, nm_female] == [0, 0]:\n",
    "#                 out.append([ind, False])\n",
    "\n",
    "#             elif (nm_male == 0) & (nm_female > 0):\n",
    "#                 out.append([ind, pass_b_female])\n",
    "\n",
    "#             elif (nm_male > 0) & (nm_female == 0):\n",
    "#                 out.append([ind, pass_b_male])\n",
    "\n",
    "#             elif (nm_male > 0) & (nm_female > 0):\n",
    "#                 if all([pass_b_male, pass_b_female]):\n",
    "#                     out.append([ind, True])\n",
    "#                 else:\n",
    "#                     perc_diff = (ndm + ndf)/(nm_male + nm_female)\n",
    "#                     corr = get_corr_at_samples(ind1, ind2, us, cns_t)\n",
    "#                     if all([(cm > 0.9), (cf > 0.9), (perc_diff <=0.2), (d < 30000), (corr > 0.9)]):\n",
    "#                         print ind\n",
    "#                         out.append([ind, True])\n",
    "#                     else:\n",
    "#                         out.append([ind,False])\n",
    "\n",
    "                    \n",
    "#             else:\n",
    "#                 print \"didn't account for all scenarios\"\n",
    "#                 break\n",
    "                \n",
    "#     df = pd.DataFrame(out, columns=['cat_pair', 'passing_consensus'])\n",
    "#     df.index = df.cat_pair\n",
    "#     return df\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
