{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import pandas as pd  \n",
    "import matplotlib.pyplot as plt  \n",
    "from ipywidgets import *  \n",
    "from IPython.display import display  \n",
    "# from IPython.html import widgets  \n",
    "# plt.style.use('ggplot')\n",
    "\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import datetime\n",
    "from subprocess import call\n",
    "import subprocess\n",
    "import glob\n",
    "import djPyBio as DJ\n",
    "import pandas as pd\n",
    "import csv\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "import copy \n",
    "import pybedtools\n",
    "import itertools \n",
    "pd.set_option('display.max_columns', 500)\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.collections import BrokenBarHCollection\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "idiogram_loc = '/frazer01/projects/CARDIPS/analysis/cardips-cnv-analysis/data/ideogram.txt'\n",
    "gs_calls_bed = '/frazer01/projects/CARDIPS/analysis/cardips-cnv-analysis/private_output/gs_calls_trunc.tsv'\n",
    "gs_calls_trunc = pd.read_table(gs_calls_bed, index_col=0)\n",
    "gs_calls_trunc['Chr']= gs_calls_trunc.Chr.apply(lambda x: 'chr' + x)\n",
    "gs_bed = pybedtools.BedTool.from_dataframe(gs_calls_trunc.iloc[:,0:4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Norm_Chroms = ['chr%s' % i for i in range(1, 23) + ['X', 'Y']]\n",
    "\n",
    "chrom_sizes = pd.read_table('/frazer01/projects/CARDIPS/analysis/cardips-cnv-analysis/data/hg19.chrom.sizes', \n",
    "                            names=['chrom', 'end'])\n",
    "\n",
    "chrom_sizes.index= chrom_sizes.chrom\n",
    "chrom_dict = chrom_sizes.ix[Norm_Chroms].end.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def chromosome_collections(df, y_positions, height,  cent_buffer=0.3, ideo=False, **kwargs):\n",
    "    \"\"\"\n",
    "    Yields BrokenBarHCollection of features that can be added to an Axes\n",
    "    object.\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pandas.DataFrame\n",
    "        Must at least have columns ['chrom', 'start', 'end', 'color']. If no\n",
    "        column 'width', it will be calculated from start/end.\n",
    "    y_positions : dict\n",
    "        Keys are chromosomes, values are y-value at which to anchor the\n",
    "        BrokenBarHCollection\n",
    "    ideo:default(False)- is this an idiogram plot? in which case, you can choose to shrink the height \n",
    "    of the centromeres\n",
    "    \n",
    "    cent_buffer:float amount to move up the centromere y_pos, with proportional decrease in\n",
    "    height to give centromere more distinctive appearance\n",
    "    \n",
    "    height : float\n",
    "        Height of each BrokenBarHCollection\n",
    "    Additional kwargs are passed to BrokenBarHCollection\n",
    "    \"\"\"\n",
    "    del_width = False\n",
    "    \n",
    "    if 'width' not in df.columns:\n",
    "        del_width = True\n",
    "        df['width'] = df['end'] - df['start']\n",
    "        \n",
    "        \n",
    "    if ideo==True:\n",
    "    \n",
    "        for chrom, group in df.groupby('chrom'):\n",
    "            \n",
    "            for stain, g in group.groupby('gieStain'):\n",
    "#                 print stain\n",
    "                yrange = ''\n",
    "                xranges =''\n",
    "                if stain == 'acen':\n",
    "                    yrange = (y_positions[chrom] + cent_buffer, height - (cent_buffer*2))\n",
    "                    xranges = g[['start', 'width']].values\n",
    "                else:\n",
    "                    yrange = (y_positions[chrom], height)\n",
    "                    xranges = g[['start', 'width']].values\n",
    "                yield BrokenBarHCollection(\n",
    "                    xranges, yrange, facecolors=g['colors'], **kwargs)\n",
    "    \n",
    "    if ideo==False:\n",
    "        \n",
    "        for chrom, group in df.groupby('chrom'):\n",
    "            yrange = (y_positions[chrom], height)\n",
    "            xranges = group[['start', 'width']].values\n",
    "            yield BrokenBarHCollection(xranges, yrange, facecolors=group['colors'], **kwargs)\n",
    "    \n",
    "    \n",
    "    if del_width:\n",
    "        del df['width']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "       \n",
    "def pos_2_df(x, chr_suff = True):\n",
    "    \n",
    "    \"\"\" make a dataframe from a ucsc coord or\n",
    "    a coordinate from GS\n",
    "    \"\"\"\n",
    "    try:\n",
    "        chrom = x.split(':')[0].replace('chr','')\n",
    "        start = int(x.split(':')[1].split('-')[0])\n",
    "        end = int(x.split(':')[1].split('-')[1])\n",
    "        name = \"POS_{}_{}_{}\".format(chrom, start, end)\n",
    "       \n",
    "    except:\n",
    "        try:\n",
    "            chrom = x.split('_')[1]\n",
    "            start = int(x.split('_')[2])\n",
    "            end = int(x.split('_')[3])\n",
    "            name = \"CNV_{}_{}_{}\".format(chrom, start, end)\n",
    "\n",
    "        except:\n",
    "            print 'enter correct format for position'\n",
    "    \n",
    "    if chr_suff:\n",
    "        data = ['chr' + chrom, start,end,name]\n",
    "        cols = ['chrom', 'start', 'end', 'name']\n",
    "        d_dict = dict(zip(cols,data))\n",
    "        df = pd.DataFrame.from_records([d_dict])\n",
    "        df = df[cols]\n",
    "        return df, 'chr' + chrom\n",
    "    if not chr_suff:\n",
    "        data = [chrom.replace('chr',''), start,end,name]\n",
    "        cols = ['chrom', 'start', 'end', 'name']\n",
    "        d_dict = dict(zip(cols,data))\n",
    "        df = pd.DataFrame.from_records([d_dict])\n",
    "        df = df[cols]\n",
    "        return df, chrom\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def calls_between_coords(x, start, end):\n",
    "    \n",
    "    \"\"\" \n",
    "    lambda function to subset calls between coordinates\n",
    "    given that columns are named Start and End for \n",
    "    Chromosomal positions\n",
    "    \n",
    "    \"\"\"\n",
    "    if x.Start > start and x.Start < end:\n",
    "        return True\n",
    "    elif x.End < end and x.End > start:\n",
    "        return True\n",
    "    \n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_hist_cov(Pos,  UUIDs, bt=False, pad=0, genome='/frazer01/publicdata/gatk_bundle_2.8/b37/human_g1k_v37_decoy_Sendai.genome', norm=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    bt = False, if true expect pybedtools BedTool Object as input for read depth prediction\n",
    "    cnv_ID = False  if true expects a cnv ID ex(CNV_CHR_START_END)\n",
    "    by default expects pos in format of UCSC Browser = chr:start-end\n",
    "    \n",
    "    POS position in format chromosome, Start, End (default chr:start-end)\n",
    "    \n",
    "    \n",
    "    UUID = iterable/list object of the samples desired\n",
    "    \n",
    "    pad = Desired amount of padding on either side of the position default=0\n",
    "    \n",
    "    genome= Desired Genome for use in padding the positions\n",
    "    \n",
    "    norm: if True, normalize read depth information to library size (under_development)\n",
    "    \n",
    "    \"\"\"\n",
    "    try:\n",
    "        \n",
    "        if bt == True:\n",
    "            bt = Pos\n",
    "        else:\n",
    "            bt = pos_2_df(Pos, chr_suff=False)[0]\n",
    "            bt = pybedtools.BedTool.from_dataframe(bt)\n",
    "\n",
    "    except:\n",
    "        print 'provide proper position and format flag'\n",
    "        \n",
    "    \n",
    "\n",
    "    if pad<>0:\n",
    "        bt = bt.slop(b=pad,g=genome)\n",
    "        \n",
    "    df_out = pd.read_table(bt.fn, names=['Chr', 'Start', 'End','ID'])\n",
    "    for ids in UUIDs:\n",
    "        \n",
    "        bw_root = '/frazer01/projects/CARDIPS/pipeline/WGS/alignment_bwa/'\n",
    "        bw_fn = bw_root+ ids + '/' + ids + '.mdup.bw'\n",
    "        Command = ['/frazer01/home/djakubosky/bin/bwtool', 'extract', 'bed', bt.fn, bw_fn,'/dev/stdout']\n",
    "        cov = subprocess.Popen(Command, stdout=subprocess.PIPE)\n",
    "        df = pd.read_table(cov.stdout, names=['Chr', 'Start', 'End','ID','Features', 'hist'])\n",
    "        df_out[ids] = df['hist'].values\n",
    "    return df_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "position = 'chr3:1230945-2340293'\n",
    "\n",
    "def plot_chrom_pos_2(stuff):\n",
    "#     try:\n",
    "#         plt.cla()\n",
    "    fig = plt.figure(figsize=(10,0.5))\n",
    "    ax = fig.add_subplot(111)\n",
    "    \n",
    "    \n",
    "    sns.set_style('ticks')\n",
    "    # Height of each ideogram\n",
    "    chrom_height = 1\n",
    "    \n",
    "    # Spacing between consecutive ideograms\n",
    "    chrom_spacing = 1\n",
    "    \n",
    "    # Height of the region track. Should be smaller than `chrom_spacing` in order to\n",
    "    # fit correctly\n",
    "    region_height = 1.5\n",
    "    \n",
    "    # Padding between the top of a region track and its corresponding ideogram\n",
    "    region_padding = 0.0\n",
    "    \n",
    "    # Decide which chromosomes to use\n",
    "    figsize=(10,0.5)\n",
    "    \n",
    "    desired = pos_2_df(stuff)[1]\n",
    "    \n",
    "    print desired\n",
    "    \n",
    "    chromosome_list = ['chr%s' % i for i in range(1, 23) + ['X', 'Y']]\n",
    "    \n",
    "    chromosome_list=[i for i in chromosome_list if i==desired]\n",
    "    \n",
    "    # Keep track of the y positions for ideograms and regions for each chromosome,\n",
    "    # and the center of each ideogram (which is where we'll put the ytick labels)\n",
    "    ybase = 10\n",
    "    chrom_ybase = {}\n",
    "    region_ybase = {}\n",
    "    chrom_centers = {}\n",
    "    \n",
    "    # Iterate in reverse so that items in the beginning of `chromosome_list` will\n",
    "    # appear at the top of the plot\n",
    "    for chrom in chromosome_list[::-1]:\n",
    "        chrom_ybase[chrom] = ybase\n",
    "        chrom_centers[chrom] = ybase + chrom_height / 2.\n",
    "        region_ybase[chrom] = ybase - 0.5 - region_padding\n",
    "        ybase += chrom_height + chrom_spacing\n",
    "    \n",
    "    # Read in ideogram.txt, downloaded from UCSC Table Browser\n",
    "    ideo = pd.read_table(\n",
    "        idiogram_loc,\n",
    "        skiprows=1,\n",
    "        names=['chrom', 'start', 'end', 'name', 'gieStain']\n",
    "    )\n",
    "    \n",
    "    # Filter out chromosomes not in our list\n",
    "    ideo = ideo[ideo.chrom.apply(lambda x: x in chromosome_list)]\n",
    "    \n",
    "    # Add a new column for width\n",
    "    ideo['width'] = ideo.end - ideo.start\n",
    "    \n",
    "    # Colors for different chromosome stains\n",
    "    color_lookup = {\n",
    "        'gneg': (1., 1., 1.),\n",
    "        'gpos25': (.6, .6, .6),\n",
    "        'gpos50': (.4, .4, .4),\n",
    "        'gpos75': (.2, .2, .2),\n",
    "        'gpos100': (0., 0., 0.),\n",
    "        'acen': (.8, .4, .4),\n",
    "        'gvar': (.8, .8, .8),\n",
    "        'stalk': (.9, .9, .9),\n",
    "    }\n",
    "    \n",
    "    # Add a new column for colors\n",
    "    ideo['colors'] = ideo['gieStain'].apply(lambda x: color_lookup[x])\n",
    "    \n",
    "    \n",
    "    # Same thing for regions\n",
    "    regions = pos_2_df(stuff)[0]\n",
    "    regions = regions[regions.chrom.apply(lambda x: x in chromosome_list)]\n",
    "    regions['width'] = regions.end - regions.start\n",
    "    regions['colors'] = '#eb6841'\n",
    "    \n",
    "    \n",
    "#     fig = plt.figure(figsize=figsize)\n",
    "#     ax = fig.add_subplot(111)\n",
    "    \n",
    "    # Now all we have to do is call our function for the ideogram data...\n",
    "    \n",
    "    for collection in chromosome_collections(ideo, chrom_ybase, chrom_height, ideo=True, linewidths=2):\n",
    "        ax.add_collection(collection)\n",
    "    \n",
    "    # Add the Marker for the location\n",
    "    \n",
    "    for collection in chromosome_collections(regions, region_ybase, region_height, alpha=1, linewidths=3, edgecolors=regions['colors']):\n",
    "        ax.add_collection(collection)\n",
    "    \n",
    "    # Axes tweaking\n",
    "    ax.set_yticks([chrom_centers[i] for i in chromosome_list])\n",
    "    ax.set_yticklabels(chromosome_list, fontsize=12)\n",
    "    \n",
    "    # labels= [int(100*float(x.get_text()))for x in ax.get_xticklabels()[:-1]]\n",
    "    # print ax.get_xlim()\n",
    "    # l = [x.get_text() for x in ax.get_xticklabels()]\n",
    "    # print l\n",
    "    \n",
    "    chrom_xticks = np.arange(0, chrom_dict[desired]+10000000, 50000000)\n",
    "    chrom_xtick_labels = [str(int(x)) + ' MB' for x in chrom_xticks/10e5]\n",
    "    \n",
    "    ax.set_xticks(chrom_xticks)\n",
    "    ax.set_xticklabels(chrom_xtick_labels)\n",
    "    \n",
    "    \n",
    "    \n",
    "    ax.axis('tight')\n",
    "    # ax.set_ylim(ax.get_ylim()[0], ax.get_ylim()[1])\n",
    "    ax.set_xlim(ax.get_xlim()[0]-1000000, ax.get_xlim()[1]+1000000)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['left'].set_color('white')\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    \n",
    "    \n",
    "    ax.get_xaxis().tick_bottom()\n",
    "    for tic in ax.yaxis.get_major_ticks():\n",
    "        tic.tick1On = tic.tick2On = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def plot_hist_2(site, genotypes, info, UUID=0, pad = 0, CNV_plot = False, regions = False):\n",
    "    \n",
    "    \n",
    "    \n",
    "    draw_chrom = plot_chrom_pos_2(site)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \"\"\"df is the hist cov df from get_hist_cov\n",
    "    site is an ID of a specific Genome Strip Call\n",
    "    uf,\n",
    "    \n",
    "    info: pandas df with minimum of ['Chr', 'Start', 'End'] columns: \n",
    "    a df of regions you would like plotted above the coverage graph\n",
    "    \n",
    "    pad= integer of padding on each side of site that you would like default=0\n",
    "    \n",
    "    CNV_plot, default: False,  if True plot regions as horizontal lines above the \n",
    "    coverage plot\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    df = get_hist_cov(site, UUID, pad=pad)\n",
    "    subplot_num = 1\n",
    "    df = df[df.ID==site]\n",
    "    ## Get CNV_Coords\n",
    "    \n",
    "    if type(UUID)==list:\n",
    "        subplot_num = len(UUID)\n",
    "    \n",
    "    fig, axes = plt.subplots(subplot_num, 1, figsize = (10, 5*subplot_num));\n",
    "    \n",
    "    start = int(site.split('_')[2])\n",
    "    end = int(site.split('_')[3])\n",
    "    \n",
    "    # Make sure this has the right format\n",
    "    Chr = 'chr' + site.split('_')[1]\n",
    "    \n",
    "    # Plot each of the histograms for each UUID\n",
    "    \n",
    "    \n",
    "    try:\n",
    "        \n",
    "        padded_start = df.Start.tolist()[0]\n",
    "        padded_end = df.End.tolist()[0]\n",
    "    except: \n",
    "        \n",
    "        padded_start = df.Start.values.tolist()[0]\n",
    "        padded_end = df.End.values.tolist()[0]\n",
    "        \n",
    "    site_location = '{0}:{1}-{2}'.format(Chr, df.Start.tolist()[0], df.End.tolist()[0])\n",
    "    Central_CNV = site\n",
    "    \n",
    "    ultimate_max_y = []\n",
    "    for i,ids in enumerate(UUID):\n",
    "        y= df[ids].values[0]\n",
    "        y = y.split(',')\n",
    "        y = np.where(np.array(y)=='NA',0.0,y)\n",
    "        y = [float(x) for x in y] \n",
    "        x = np.arange(df.Start,df.End, 1)\n",
    "        axes[i].plot(x,y);\n",
    "        axes[i].fill_between(x,y);\n",
    "        axes[i].set_title(ids);\n",
    "        max_y = np.max(np.array(y))\n",
    "        ultimate_max_y.append(max_y)\n",
    "        \n",
    "        mean_y = np.mean(np.array(y))\n",
    "        median_y = np.median(np.array(y))\n",
    "        min_y = np.min(np.array(y))\n",
    "\n",
    "        formatted_Vals = [\"{0:.2f}\".format(l) for l in [mean_y, median_y, min_y, max_y]]\n",
    "        formatted_Vals = formatted_Vals + [site_location]\n",
    "\n",
    "        axes[i].set_ylim(0, max_y+20)\n",
    "        axes[i].set_xlim(padded_start,padded_end)\n",
    "        axes[i].text(padded_end + 1000, max_y, \"Mean: {0}\\nMedian: {1} \\nMin: {2}\\nMax: {3}\\n{4}\".format(*formatted_Vals), \n",
    "                     style='italic', horizontalalignment='left',\n",
    "                     verticalalignment='top', bbox={'facecolor':'red', 'alpha':0.5, 'pad':10})\n",
    "    \n",
    "        \n",
    "        \n",
    "        \n",
    "        if CNV_plot==True:\n",
    "            chrom_CNVS = info[info['Chr']== Chr]    \n",
    "            contained = chrom_CNVS[chrom_CNVS.apply(lambda b: calls_between_coords(b,df.Start.tolist()[0], df.End.tolist()[0]), axis=1)==True]\n",
    "\n",
    "            # Jitter the height of the hlines\n",
    "            height = [6,8] *100\n",
    "            count = 0\n",
    "\n",
    "            for s,e,f in zip(contained.Start, contained.End, contained.Coords):\n",
    "                genotype = genotypes.loc[f][ids]\n",
    "\n",
    "\n",
    "                axes[i].text(((e-s)/2)+ s,max_y + height[count]+1,genotype, horizontalalignment='center');\n",
    "                if int(genotype) < 2:\n",
    "\n",
    "                    axes[i].hlines(max_y + height[count], s,e, colors='r', alpha=0.8);\n",
    "\n",
    "                elif int(genotype) > 2:\n",
    "                    axes[i].hlines(max_y + height[count], s,e, colors='g', alpha=0.8);\n",
    "                else:\n",
    "                    axes[i].hlines(max_y + height[count], s,e, colors='black', alpha=0.8);\n",
    "                count +=1\n",
    "        \n",
    "        axes[i].vlines(start, 0, max_y, linestyles='dashed')\n",
    "        axes[i].vlines(end, 0, max_y, linestyles='dashed')\n",
    "\n",
    "        \n",
    "    \n",
    "    plt.suptitle(Central_CNV, fontsize=20);\n",
    "    \n",
    "    for i,ids in enumerate(UUID):\n",
    "        axes[i].set_ylim(0, (max(ultimate_max_y) + max(ultimate_max_y)* 0.2))\n",
    "    \n",
    "    \n",
    "    axes[len(UUID)-1].set_xlabel('Coordinates')\n",
    "    plt.show()\n",
    "    \n",
    "    if CNV_plot==True:\n",
    "        \n",
    "        display(contained);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gs_genotypes= pd.read_pickle('/frazer01/projects/CARDIPS/analysis/cardips-cnv-analysis/private_output/cnv_processing/gs_cns.pkl')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
